game/system/IOP_Kernel.cpp

```diff
@@ -118,6 +133,38 @@ s32 IOP_Kernel::WaitSema(s32 id) {
   return KE_OK;
 }
 
+s32 IOP_Kernel::ReceiveMbx(void** msg, s32 id) {
+  auto& box = mbxs.at(id);
+  if (!box.messages.empty()) {
+    auto ret = PollMbx(msg, id);
+    ASSERT(ret == KE_OK);
+    return KE_OK;
+  }
+
+  ASSERT(!box.wait_thread);  // don't know how to deal with this, hopefully doesn't come up.
```

Comment:
The real kernel would have a linked list of waiting threads sorted in either FIFO or priority order depending on mailbox settings.